contract/src/main.rs

```diff
@@ -568,16 +586,32 @@ pub extern "C" fn mint() {
         }
     }
 
-    // The contract's ownership behavior (determined at installation) determines,
-    // who owns the NFT we are about to mint.()
-    let ownership_mode = utils::get_ownership_mode().unwrap_or_revert();
-    let caller = utils::get_verified_caller().unwrap_or_revert();
-    let token_owner_key: Key =
-        if let OwnershipMode::Assigned | OwnershipMode::Transferable = ownership_mode {
-            runtime::get_named_arg(ARG_TOKEN_OWNER)
-        } else {
-            caller
-        };
+    // Revert if minting is acl and caller is not whitelisted.
+    if MintingMode::ACL == minting_mode {
+        let is_whitelisted = utils::get_dictionary_value_from_key::<bool>(
+            ACL_WHITELIST,
+            &utils::encode_dictionary_item_key(caller),
+        )
+        .unwrap_or_default();
+        if !is_whitelisted {
+            match caller.tag() {
+                KeyTag::Hash => {
+                    if let Some(contract_package) = contract_package {
+                        let is_package_whitelisted = utils::get_dictionary_value_from_key::<bool>(
+                            ACL_WHITELIST,
+                            &utils::encode_dictionary_item_key(contract_package),
+                        )
+                        .unwrap_or_default();
+                        if !is_package_whitelisted {
+                            runtime::revert(NFTCoreError::UnlistedContractHash)
+                        }
+                    }
+                }
+                KeyTag::Account => runtime::revert(NFTCoreError::InvalidMinter),
+                _ => runtime::revert(NFTCoreError::InvalidKey),
+            }
+        }
```

Comment:
There is probably a version possible without intermediary booleans, is_whitelisted is_package_whitelisted that would reduce the number of lines but to me the logic becomes more obscure to read for a human, I like to understand just by reading the if statement why a reverts happens without the need to read lines above.

Reply:
will get back to you about this at tomorrow

Reply:
I changed the logic to test the contract package first as maybe this would save some cost for contracts check. Let me know if this looks better to you or let me know otherwise thanks.